046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,93 @@
+
+The current flow of KafkaRoller
+![The current flow of KafkaRoller](images/046-kafka-roller-current-flow.png)
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller could end up force restarting the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+
+KafkaAgent currently collects the `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is greater than `3`(`RUNNING`) and not `127` (`UNKNOWN`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation
+
+To address the issues with restarts and log recovery, KafkaRoller needs to be aware of the current broker state.  If a broker is performing log recovery, KafkaRoller should fail the reconciliation instead of force restarting it.
+
+## Proposal
+ 
+The objective of this proposal is to introduce a REST API endpoint within KafkaAgent, which exposes metrics from a Kafka broker. Furthermore, KafkaRoller would be modified so that it can query this endpoint to determine the current state of a broker, thus preventing unnecessary restarts. Additionally, KafkaRoller would allow a human operator to observe whether the log recovery process is making progress or is stuck.
+
+The following implementation details will describe behaviour of the new API in KafkaAgent and how KafkaRoller would interact with it and handle a broker that is in log recovery.
+
+### Implementation details
+
+**KafkaAgent**:
+
+- Add a web server to accept SSL connections on port 8443.
+- Use Jetty for the web server since Kafka already uses it therefore avoids an additional dependency.
+- Reuse the broker's own certificate that would be in the container for TLS authentication. 
+- Configure the server to require TLS client authentication. 
+- Expose an endpoint `v1/recovery-state` that includes the API version in the URI path.
+- Extend KafkaAgent to collect `kafka.log:type=LogManager,name=remainingLogsToRecover` and `kafka.log:type=LogManager,name=remainingSegmentsToRecover` metrics.
+- Use Jetty's native ServerHandler API to handle a `GET` request and return a JSON response that looks like this:
+```
+{
+  "brokerState": 2,
+  "remainingLogsToRecover": 123,
+  "remainingSegmentsToRecover": 456
+}
+```
+- Return HTTP 404 (Not Found) if a request is sent to an unknown version of the API e.g. `GET /v5/recovery-state` and `v5` does not exist.
+
+**KafkaRoller**:
+
+The proposed flow of KafkaRoller
+![The proposed flow of KafkaRoller](images/046-kafka-roller-new-flow.png)
+
+If a kafka pod does not become ready within the operational timeout (5 mins by default) take the following actions:
+1. Check if the maximum number of retries has been reached. If it has, fail the reconciliation.
```

Comment:
How many retries do we have for the Pod readiness here? This seems to suggest we wait the 5 minutes repeatedly here in one reconciliation? Or does it men something else?

Reply:
The default max attempt is 6. We would be following the current behaviour here, which waits for 5 minutes for pod readiness and then retry as long as we haven't hit the max attempts. So yes it already repeatedly waits for 5 minutes up to 6 times.

Reply:
I guess it makes sense to stick with what we have in this proposal. But it might be worth clarifying in the proposal.